### ISSUE TITLE
Which links to select for get_lagged_dependencies? (Warning: Link specified in selected links that is outside the scope of the selected variables)

### ISSUE THREAD

**Issue by shaypal5** [jakobrunge/tigramite#35]
Hey Jakob and other contributors! :)

So I'm using PCMCI to find causal relations among a set of 13 variables, indexed from 0 to 12. However, I'm only interested in the causal parents of variables 5 to 12, so I initialize `PCMCI` with `selected_variables=[0,...,12]`.

However, when running the `get_lagged_dependencies()` method of the initialized `PCMCI` object, I only supply `tau_max`, leaving `selected_links` with its default `None`, so all links are tested (as per the documentation).

This is intentional, as I figured that even though I'm only interested in the parents of some of the variables, I should take into consideration all variable-pair interactions when choosing the max lag to use, as they might be crucial for the correct estimation of any causal link in the system.

However, the `get_lagged_dependencies` method throws the following warning (coming from the `_set_sel_links()` input validation method):

```python
"Warning: Link specified in selected links that is outside the scope of the selected variables"
```

As a result, I wanted to ask if my logic is correct, and if no, how to construct a correct set of selected links. Is it simply the set of all links who has at least one member in `selected_variables`?

Also, is there a chance this is what causing the error I have when running this method (see [issue #33](https://github.com/jakobrunge/tigramite/issues/33)?

Thank you again for all your help,
Shay

**Comment by jakobrunge** [comment]
Dear Shay
The pcmci.get_lagged_dependencies method does NOT run the pcmci algorithm and will only output correlations (with  ParCorr as an instantiated independence test). You can set parents as an optional parameter, but it will not iteratively identify the underlying causal graph.

Nevertheless, if you run what you wrote above with the run_pcmci function, you will get what you are looking for. you can ignore the warning, it will identify the parents of the selected variables conditional on any other variable, since you did not restrict the selected links.

**Comment by shaypal5** [comment]
Hey Jakob,

Yes, of course, I know `get_lagged_dependencies` does not run PCMCI. :)

I am using it to find an appropriate value for `tau_max` for `run_pcmci` in a semi-automatic way, as I have discussed in issue #32 . This is of course based on the manual way to estimate a proper value for `tau_max` based on the output of `get_lagged_dependencies`, demonstrated in your [basics tutorial](https://github.com/jakobrunge/tigramite/blob/master/tutorials/tigramite_tutorial_basics.ipynb).

When I wrote that "I'm using PCMCI to find causal relations " I meant generally, in one of my current projects. I'm attaching the full flow below (also found in the code + data attachment I provided to help debug issue #33 ).

Now that you can rest assured that I understand what `get_lagged_dependencies` and `run_pcmci` are meant to do, and that you see the way I use them, I'd love to get an answer to my original question:

> Is it true that even though I'm only interested in the parents of some of the variables, I should take into consideration all variable-pair interactions when choosing the max lag to use, as they might be crucial for the correct estimation of any causal link in the system?

Full code:

```python
"""Reproducig some tigramite error."""

import math

import numpy as np
import pandas as pd

from tigramite.pcmci import PCMCI
from tigramite.independence_tests import CMIsymb
from tigramite.data_processing import DataFrame


SECONDS_IN_DAY = 60*60*24
DEF_CORRELATION_EPSILON = 0.25


def series_max_lag(arr, epsilon=None):
    if epsilon is None:
        epsilon = DEF_CORRELATION_EPSILON
    in_epsilon_range = (-epsilon < arr) & (arr < epsilon)
    if any(in_epsilon_range):
        return np.argmax(in_epsilon_range)
    return len(arr)


def find_max_lag(pcmci, timestep_length_in_sec, epsilon=None):
    timesteps_in_a_day = math.ceil(SECONDS_IN_DAY / timestep_length_in_sec)
    correlations = pcmci.get_lagged_dependencies(tau_max=timesteps_in_a_day)
    return np.max(np.apply_along_axis(series_max_lag, 2, correlations))


def example_problem():
    df = pd.read_csv('final_input_example.csv')
    print((f"Data min: {df.min().min()}, max: {df.max().max()}, "
           f"sum: {df.sum().sum()}"))
    data = df.values

    # building a tigramite dataframe object
    tdf = DataFrame(
        data=data,
        mask=None,
        missing_flag=None,
        var_names=df.columns,
        datatime=None,
    )

    # initialize independence test object and PCMCI object
    indp_test = CMIsymb(
        n_symbs=None,
        significance=None,
        sig_blocklength=None,
        conf_blocklength=None,
    )
    selected_variables = [
        col_lbl for col_lbl in df.columns
        if 'i' in col_lbl
    ]
    selected_variables_ix = [
        df.columns.get_loc(lbl) for lbl in selected_variables]
    print((
        "Init PCMCI with:"
        f"dataframe={tdf},"
        f"cond_ind_test={indp_test},"
        f"selected_variables={selected_variables_ix},"
        f"verbosity=10,"
    ))
    pcmci = PCMCI(
        dataframe=tdf,
        cond_ind_test=indp_test,
        selected_variables=selected_variables_ix,
        verbosity=10,
    )

    # finding the max lag to use
    print("Finding max lag to use...")
    timestep_in_sec = 600
    max_lag = find_max_lag(
        pcmci=pcmci,
        timestep_length_in_sec=timestep_in_sec,
        epsilon=None,
    )

    # running pcmci
    print("Running PCMCI...")
    results = pcmci.run_pcmci(tau_max=max_lag, pc_alpha=None)

    print("Done!")
    print(results)


if __name__ == '__main__':
    example_problem()
```

**Comment by jakobrunge** [comment]
I hope my reply and bug fix in the developer fixes this issue?

**Comment by shaypal5** [comment]
Hmmm. Not really.

As I've said, when running the `get_lagged_dependencies()` method of the initialized PCMCI object, I only supply `tau_max`, leaving `selected_links` with its default `None`, so all links are tested (as per the documentation).

This is intentional, as I figured that even though I'm only interested in the parents of some of the variables, I should take into consideration all variable-pair interactions when choosing the max lag to use, as they might be crucial for the correct estimation of any causal link in the system.

So I have two questions here:
1. When using `get_lagged_dependencies()` to get an idea about the right value for `tau_max` to use later for `run_pcmci()` (whether this is done manually or using a method like in issue #32 is besides the point), **should all links be tested, even though we would later want to infer causal parents for only a subset of the provided variables in the system?**
2. If the answer to the first answer is yes, does setting `selected_links=None` achieves this (testing all possible causal links)? Your answer to closed issue #37 actually suggests that **setting `selected_links=None` will not achieves this!**

**Comment by jakobrunge** [comment]
ad 1) No, you only need to choose the maximum lag of a parent relevant to your selected_variables. But you should set selected_links=None to look for links from ALL other variables. With one (quite important) caveat, the problem of inferring lagged parents is completely independent for each selected_variable.

Here's the caveat. The idea of the MCI test in PCMCI is to condition on the parents of both the driven variable (at time t) and the parents of the lagged variables (at time t-tau). The advantage of this is that autocorrelation is better accounted for and you get proper control of false positives. Of course, this requires that you estimate the parents of ALL variables (selected_variables=None). Now that might be very time-consuming. If you restrict selected_variables to a subset, then the MCI test will only condition on parents of that subset. 
Here's a workaround: You can infer the parents of your desired selected_variables using run_pc_stable(). Do not restrict selected_links. Then you ADD to these parents the autocorrelations also of those variables NOT included in selected_variables. I.e.
`for j in range(N):
    if (j, -1) not in all_parents[j]: all_parents[j].append( (j, -1) )`
That alleviates the problem of false positives.

In general, you would use selected_links only to contrain some lags, e.g., if you know that some links don't exist as prior knowledge...
